packages/babel-parser/src/parser/expression.js

```diff
@@ -2186,7 +2191,9 @@ export default class ExpressionParser extends LValParser {
   isAwaitAllowed(): boolean {
     if (this.scope.inFunction) return this.scope.inAsync;
     if (this.options.allowAwaitOutsideFunction) return true;
-    if (this.hasPlugin("topLevelAwait")) return this.inModule;
+    if (this.hasPlugin("topLevelAwait")) {
+      return this.inModule && this.scope.inAsync;
```

Comment:
Since at this point `this.scope.inFunction` is false, doesn't `this.inModule` already imply `this.scope.inAsync`?

Reply:
Nope. For example
```js
class C {
  p = await 0;
}
```
Even top level await is enabled, await cant not be allowed here because it is _not_ allowed in field initializers.

I think the idea of top level await is that the whole module is wrapped in an async function, which is what I do here to assign `SCOPE_ASYNC` to the program scope. Then we will need to handle cases where `await` is _not_ allowed.